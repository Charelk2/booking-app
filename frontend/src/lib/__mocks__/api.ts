const handler: ProxyHandler<any> = {
  get: (target, prop) => {
    if (!(prop in target)) {
      target[prop] = jest.fn();
    }
    return target[prop];
  },
};

const api: any = new Proxy(
  { getServiceCategories: jest.fn().mockResolvedValue({ data: [] }) },
  handler,
);

module.exports = api;
