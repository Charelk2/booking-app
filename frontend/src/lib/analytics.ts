export type AnalyticsPayload = Record<string, unknown>;

export const trackEvent = (event: string, payload: AnalyticsPayload = {}): void => {
  if (typeof window !== 'undefined') {
    const anyWindow = window as any;
    if (anyWindow.analytics && typeof anyWindow.analytics.track === 'function') {
      anyWindow.analytics.track(event, payload);
      return;
    }
  }
  // Fallback to console so events are still visible during development
  // eslint-disable-next-line no-console
  console.info('analytics event', event, payload);
};
