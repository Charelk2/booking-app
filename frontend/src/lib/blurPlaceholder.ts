export const BLUR_PLACEHOLDER =
  'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMSIgaGVpZ2h0PSIxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxyZWN0IHdpZHRoPSIxIiBoZWlnaHQ9IjEiIGZpbGw9IiNlZWVlZWUiIC8+PC9zdmc+';
