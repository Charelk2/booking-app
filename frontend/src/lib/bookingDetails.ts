export interface ParsedBookingDetails {
  eventType?: string;
  description?: string;
  date?: string;
  location?: string;
  guests?: string;
  venueType?: string;
  soundNeeded?: string;
  notes?: string;
}

import { BOOKING_DETAILS_PREFIX } from './constants';

export function parseBookingDetailsFromMessage(content: string): ParsedBookingDetails {
  const details: ParsedBookingDetails = {};
  const lines = content.replace(BOOKING_DETAILS_PREFIX, '').trim().split('\n');
  lines.forEach((line) => {
    const parts = line.split(':');
    if (parts.length >= 2) {
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      switch (key) {
        case 'Event Type':
          details.eventType = value;
          break;
        case 'Description':
          details.description = value;
          break;
        case 'Date':
          details.date = value;
          break;
        case 'Location':
          details.location = value;
          break;
        case 'Guests':
          details.guests = value;
          break;
        case 'Venue':
          details.venueType = value;
          break;
        case 'Sound':
          details.soundNeeded = value;
          break;
        case 'Notes':
          details.notes = value;
          break;
        default:
          break;
      }
    }
  });
  return details;
}
