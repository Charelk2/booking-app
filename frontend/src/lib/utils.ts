import api from './api';

export const getFullImageUrl = (relativePath: string | undefined | null): string | null => {
  if (!relativePath) return null;
  if (relativePath.startsWith('http://') || relativePath.startsWith('https://')) {
    return relativePath;
  }
  const cleanPath = relativePath.startsWith('/static/') ? relativePath : `/static/${relativePath.replace(/^\/+/, '')}`;
  return `${api.defaults.baseURL}${cleanPath}`;
};

export const extractErrorMessage = (detail: unknown): string => {
  if (!detail) return 'An unexpected error occurred.';
  if (typeof detail === 'string') return detail;
  if (Array.isArray(detail)) {
    return detail
      .map((d) => {
        if (typeof d === 'object' && d && 'msg' in d) {
          const record = d as Record<string, unknown>;
          return record.msg as string;
        }
        return JSON.stringify(d);
      })
      .join(', ');
  }
  if (typeof detail === 'object') {
    const record = detail as Record<string, unknown>;
    if (typeof record.msg === 'string') return record.msg;
    return JSON.stringify(detail);
  }
  return String(detail);
};
