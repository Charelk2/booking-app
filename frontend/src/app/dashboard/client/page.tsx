"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

