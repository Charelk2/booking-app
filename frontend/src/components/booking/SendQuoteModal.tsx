import React, { useState, useEffect, useRef } from 'react';
import { format } from 'date-fns';
import Button from '../ui/Button';
import { ServiceItem, QuoteV2Create, QuoteTemplate } from '@/types';
import { getQuoteTemplates } from '@/lib/api';
import { formatCurrency, generateQuoteNumber } from '@/lib/utils';

interface Props {
  open: boolean;
  onClose: () => void;
  onSubmit: (data: QuoteV2Create) => Promise<void> | void;
  artistId: number;
  clientId: number;
  bookingRequestId: number;
  serviceName?: string;
  // New props for initial quote data
  initialBaseFee?: number;
  initialTravelCost?: number;
  initialSoundNeeded?: boolean; // New prop
}

const expiryOptions = [
  { label: '1 day', value: 24 },
  { label: '3 days', value: 72 },
  { label: '7 days', value: 168 },
];

const SendQuoteModal: React.FC<Props> = ({
  open,
  onClose,
  onSubmit,
  artistId,
  clientId,
  bookingRequestId,
  serviceName,
  initialBaseFee,
  initialTravelCost,
  initialSoundNeeded, // Destructure new prop
}) => {
  const [services, setServices] = useState<ServiceItem[]>([]);
  const [serviceFee, setServiceFee] = useState(0);
  const [soundFee, setSoundFee] = useState(0);
  const [travelFee, setTravelFee] = useState(0);
  const [accommodation, setAccommodation] = useState('');
  const [discount, setDiscount] = useState(0);
  const [expiresHours, setExpiresHours] = useState<number | null>(null);
  const [templates, setTemplates] = useState<QuoteTemplate[]>([]);
  const [selectedTemplate, setSelectedTemplate] = useState<number | ''>('');
  const [quoteNumber, setQuoteNumber] = useState('');
  const [description, setDescription] = useState('');

  const currentDate = format(new Date(), 'PPP');

  // Ref to track if we already applied the initial values so we don't
  // clobber user edits when props update
  const hasPrefilled = useRef(false);

  useEffect(() => {
    if (!open) {
      // reset when modal closes so next open can prefill again
      hasPrefilled.current = false;
      setSelectedTemplate('');
      return;
    }

    // fetch templates and generate quote number each time the modal opens
    getQuoteTemplates(artistId)
      .then((res) => setTemplates(res.data))
      .catch(() => setTemplates([]));
    setQuoteNumber(generateQuoteNumber());

    if (selectedTemplate === '') {
      // always clear optional fields when no template is selected
      setServices([]);
      setAccommodation('');
      setDiscount(0);
      setExpiresHours(null);
      setDescription('');

      // apply initial props once when they become available
      if (!hasPrefilled.current &&
          (typeof initialBaseFee === 'number' ||
            typeof initialTravelCost === 'number' ||
            typeof initialSoundNeeded === 'boolean')) {
        setServiceFee(typeof initialBaseFee === 'number' ? initialBaseFee : 0);
        setTravelFee(typeof initialTravelCost === 'number' ? initialTravelCost : 0);
        setSoundFee(initialSoundNeeded ? 250 : 0);
        hasPrefilled.current = true;
      }
    }
  }, [open, selectedTemplate, artistId, initialBaseFee, initialTravelCost, initialSoundNeeded]);

  useEffect(() => {
    const tmpl = templates.find((t) => t.id === selectedTemplate);
    if (tmpl) {
      // Apply template values
      setServiceFee(Number(tmpl.services[0]?.price || 0));
      setServices(tmpl.services.slice(1));
      setSoundFee(tmpl.sound_fee);
      setTravelFee(tmpl.travel_fee);
      setAccommodation(tmpl.accommodation || '');
      setDiscount(tmpl.discount || 0);
    } else if (selectedTemplate === '') {
      // If "Choose template" is selected (or no template was initially chosen),
      // revert to initial props or default to 0/empty
      setServiceFee(typeof initialBaseFee === 'number' ? initialBaseFee : 0);
      setTravelFee(typeof initialTravelCost === 'number' ? initialTravelCost : 0);
      setSoundFee(initialSoundNeeded ? 250 : 0); // Revert soundFee based on initialSoundNeeded
      setAccommodation('');
      setDiscount(0);
      setServices([]);
    }
  }, [selectedTemplate, templates, initialBaseFee, initialTravelCost, initialSoundNeeded]); // Added initialSoundNeeded to dependencies

  const subtotal =
    serviceFee + services.reduce((acc, s) => acc + Number(s.price), 0) + soundFee + travelFee;
  const subtotalAfterDiscount = subtotal - (discount || 0);
  const taxesFees = subtotalAfterDiscount * 0.15;
  const estimatedTotal = subtotalAfterDiscount + taxesFees;

  const addService = () => setServices([...services, { description: '', price: 0 }]);
  const removeService = (idx: number) =>
    setServices(services.filter((_, i) => i !== idx));
  const updateService = (idx: number, field: keyof ServiceItem, value: string) => {
    const updated = services.map((s, i) => (i === idx ? { ...s, [field]: field === 'price' ? Number(value) : value } : s));
    setServices(updated);
  };

  const handleSubmit = async () => {
    const expires_at = expiresHours ? new Date(Date.now() + expiresHours * 3600000).toISOString() : null;
    await onSubmit({
      booking_request_id: bookingRequestId,
      artist_id: artistId,
      client_id: clientId,
      services: [
        { description: serviceName ?? 'Service fee', price: serviceFee },
        ...services,
      ],
      sound_fee: soundFee,
      travel_fee: travelFee,
      accommodation: accommodation || null,
      discount: discount || null,
      expires_at,
    });
  };

  if (!open) return null;
  return (
    <div className="fixed inset-0 bg-black/50 flex items-center justify-center z-50 p-4">
      <div className="bg-white rounded-xl shadow-2xl w-full max-w-md p-6 border border-gray-100">
        <div className="flex items-center justify-between mb-4">
          <h2 className="text-xl font-semibold text-gray-800">View Quote</h2>
          {templates.length > 0 && (
            <select
              className="border border-gray-300 rounded-md px-3 py-1.5 text-sm focus:outline-none focus:ring-2 focus:ring-indigo-500 transition-colors"
              value={selectedTemplate}
              onChange={(e) =>
                setSelectedTemplate(e.target.value ? Number(e.target.value) : '')
              }
            >
              <option value="">Choose template</option>
              {templates.map((t) => (
                <option key={t.id} value={t.id}>
                  {t.name}
                </option>
              ))}
            </select>
          )}
        </div>
        <div className="space-y-4 max-h-[65vh] overflow-y-auto pr-2">
          <div className="flex flex-col gap-y-2 mb-2 text-sm text-gray-600">
            <div><span className="font-medium">Quote No:</span> {quoteNumber}</div>
            <div><span className="font-medium">Date:</span> {currentDate}</div>
            <input
              type="text"
              className="border border-gray-300 rounded-md p-2 w-full focus:outline-none focus:ring-2 focus:ring-indigo-500"
              placeholder="Quote Description (e.g., 'Wedding Performance')"
              value={description}
              onChange={(e) => setDescription(e.target.value)}
            />
          </div>

          <div className="space-y-3">
            <label htmlFor="service-fee" className="flex items-center gap-2 text-sm font-normal border border-gray-200 rounded-lg p-3 bg-gray-50">
              <span className="flex-1 font-medium text-gray-700">{serviceName ?? 'Service'} fee</span>
              <input
                id="service-fee"
                type="number"
                inputMode="numeric"
                className="w-28 border border-gray-300 rounded-md p-1.5 text-right focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="0.00"
                value={serviceFee}
                onChange={(e) => setServiceFee(Number(e.target.value))}
              />
            </label>
            <label htmlFor="sound-fee" className="flex items-center gap-2 text-sm font-normal border border-gray-200 rounded-lg p-3 bg-gray-50">
              <span className="flex-1 font-medium text-gray-700">Sound fee</span>
              <input
                id="sound-fee"
                type="number"
                inputMode="numeric"
                className="w-28 border border-gray-300 rounded-md p-1.5 text-right focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="0.00"
                value={soundFee}
                onChange={(e) => setSoundFee(Number(e.target.value))}
              />
            </label>
            <label htmlFor="travel-fee" className="flex items-center gap-2 text-sm font-normal border border-gray-200 rounded-lg p-3 bg-gray-50">
              <span className="flex-1 font-medium text-gray-700">Travel fee</span>
              <input
                id="travel-fee"
                type="number"
                inputMode="numeric"
                className="w-28 border border-gray-300 rounded-md p-1.5 text-right focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="0.00"
                value={travelFee}
                onChange={(e) => setTravelFee(Number(e.target.value))}
              />
            </label>
          </div>

          {services.map((s, i) => (
            <div key={i} className="flex items-center gap-2 text-sm font-normal border border-gray-200 rounded-lg p-3 bg-gray-50">
              <input
                type="text"
                className="flex-1 border border-gray-300 rounded-md p-1.5 focus:outline-none focus:ring-2 focus:ring-indigo-500"
                placeholder="Item Description"
                value={s.description}
                onChange={(e) => updateService(i, 'description', e.target.value)}
              />
              <input
                type="number"
                className="w-28 border border-gray-300 rounded-md p-1.5 text-right focus:outline-none focus:ring-2 focus:ring-indigo-500"
                inputMode="numeric"
                placeholder="0.00"
                value={s.price}
                onChange={(e) => updateService(i, 'price', e.target.value)}
              />
              <button
                type="button"
                onClick={() => removeService(i)}
                aria-label="Remove item"
                className="text-red-500 hover:text-red-700 transition-colors text-lg font-bold"
              >
                &times;
              </button>
            </div>
          ))}
          <Button type="button" onClick={addService} className="text-sm w-full py-2.5" variant="secondary">
            + Add Custom Item
          </Button>

          <label htmlFor="accommodation" className="flex flex-col text-sm font-normal mt-4">
            <span className="font-medium text-gray-700 mb-1">Accommodation (optional)</span>
            <textarea
              id="accommodation"
              className="w-full border border-gray-300 rounded-md p-2 focus:outline-none focus:ring-2 focus:ring-indigo-500"
              placeholder="E.g., '1 night hotel stay: $150'"
              value={accommodation}
              onChange={(e) => setAccommodation(e.target.value)}
              rows={2}
            />
          </label>
          <label htmlFor="discount" className="flex items-center gap-2 text-sm font-normal mt-4 border border-gray-200 rounded-lg p-3 bg-gray-50">
            <span className="flex-1 font-medium text-gray-700">Discount (optional)</span>
            <input
              id="discount"
              type="number"
              inputMode="numeric"
              className="w-28 border border-gray-300 rounded-md p-1.5 text-right focus:outline-none focus:ring-2 focus:ring-indigo-500"
              placeholder="0.00"
              value={discount}
              onChange={(e) => setDiscount(Number(e.target.value))}
            />
          </label>
          <label htmlFor="expires-hours" className="flex flex-col text-sm font-normal mt-4">
            <span className="font-medium text-gray-700 mb-1">Expires in</span>
            <select
              id="expires-hours"
              className="w-full border border-gray-300 rounded-md p-2 focus:outline-none focus:ring-2 focus:ring-indigo-500"
              value={expiresHours ?? ''}
              onChange={(e) => setExpiresHours(e.target.value ? Number(e.target.value) : null)}
            >
              <option value="">No expiry</option>
              {expiryOptions.map((o) => (
                <option key={o.value} value={o.value}>
                  {o.label}
                </option>
              ))}
            </select>
          </label>
          <hr className="border-t border-gray-200 my-4" />
          <h5 className="font-bold text-lg mb-3 text-gray-800">Estimated Cost</h5>
          <div className="space-y-2 text-gray-700">
            <div className="flex justify-between items-center">
              <span>Artist Base Fee</span>
              <span>{formatCurrency(serviceFee)}</span>
            </div>
            <div className="flex justify-between items-center">
              <span className="flex items-center group">
                Travel
                <span className="has-tooltip relative ml-1.5">
                  <span className="cursor-pointer text-black-600">ⓘ</span>
                  <div className="tooltip absolute bottom-full mb-2 w-48 bg-gray-800 text-white text-xs rounded-md p-2 text-center z-10 hidden group-hover:block">
                    Travel costs you set for this quote.
                  </div>
                </span>
              </span>
              <span>{formatCurrency(travelFee)}</span>
            </div>
            {soundFee !== 0 && (
              <div className="flex items-center justify-between">
                <span className="flex items-center group">
                  Sound Equipment
                  <span className="has-tooltip relative ml-1.5">
                    <span className="cursor-pointer text-black-600">ⓘ</span>
                    <div className="tooltip absolute bottom-full mb-2 w-48 bg-gray-800 text-white text-xs rounded-md p-2 text-center z-10 hidden group-hover:block">
                      Standard package for events up to 150 guests.
                    </div>
                  </span>
                </span>
                <span>{formatCurrency(soundFee)}</span>
              </div>
            )}
            {discount !== 0 && (
              <div className="flex justify-between items-center">
                <span>Discount</span>
                <span>-{formatCurrency(discount)}</span>
              </div>
            )}
            <div className="flex justify-between items-center border-t pt-2 mt-2 border-dashed">
              <span className="font-medium">Subtotal</span>
              <span className="font-medium">{formatCurrency(subtotalAfterDiscount)}</span>
            </div>
            <div className="flex justify-between items-center">
              <span>Taxes & Fees (Est.)</span>
              <span>{formatCurrency(taxesFees)}</span>
            </div>
            <div className="flex justify-between items-center text-xl font-bold text-gray-900 border-t pt-3 mt-3">
              <span>Estimated Total</span>
              <span>{formatCurrency(estimatedTotal)}</span>
            </div>
          </div>
          <div className="flex items-start space-x-3 mt-6">
            <input type="checkbox" id="terms" className="mt-1 h-4 w-4 text-red-600 border-gray-300 rounded focus:ring-red-500" />
            <label htmlFor="terms" className="text-sm text-gray-600">
              I have reviewed my details and agree to the{' '}
              <a href="#" className="text-red-600 hover:underline">
                terms of service
              </a>
              .
            </label>
          </div>
        </div>
        <div className="flex gap-3 mt-6">
          <Button type="button" variant="secondary" onClick={onClose} className="flex-1 py-2.5 rounded-full border-gray-300 text-gray-700 hover:bg-gray-100">
            Cancel
          </Button>
          <Button type="button" onClick={handleSubmit} title="This quote will be sent to the client" className="flex-1 py-2.5 rounded-full bg-indigo-600 hover:bg-indigo-700 text-white font-semibold shadow-md">
            Submit Request
          </Button>
        </div>
      </div>
    </div>
  );
};

export default SendQuoteModal;