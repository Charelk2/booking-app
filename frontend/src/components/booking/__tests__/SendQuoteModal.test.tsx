import { createRoot } from 'react-dom/client';
import React from 'react';
import { act } from 'react';
import SendQuoteModal from '../SendQuoteModal';
import * as api from '@/lib/api';
import { formatCurrency } from '@/lib/utils';

jest.mock('@/lib/api');

describe('SendQuoteModal', () => {
  it('loads templates and applies selection', async () => {
    (api.getQuoteTemplates as jest.Mock).mockResolvedValue({
      data: [
        {
          id: 1,
          artist_id: 2,
          name: 'Base',
          services: [{ description: 'A', price: 5 }],
          sound_fee: 1,
          travel_fee: 2,
          accommodation: null,
          discount: null,
          created_at: '',
          updated_at: '',
        },
      ],
    });
    const div = document.createElement('div');
    const root = createRoot(div);
    await act(async () => {
      root.render(
        <SendQuoteModal
          open
          onClose={() => {}}
          onSubmit={() => {}}
          artistId={2}
          clientId={3}
          bookingRequestId={4}
        />,
      );
    });
    const select = div.querySelector('select') as HTMLSelectElement;
    expect(select).not.toBeNull();
    await act(async () => {
      select.value = '1';
      select.dispatchEvent(new Event('change', { bubbles: true }));
    });
    const inputs = div.querySelectorAll('input[type="number"]');
    expect(inputs[0].value).toBe('5');
    expect(inputs[1].value).toBe('1');
    expect(inputs[2].value).toBe('2');
    root.unmount();
  });

  it('shows formatted totals', async () => {
    (api.getQuoteTemplates as jest.Mock).mockResolvedValue({
      data: [
        {
          id: 1,
          artist_id: 2,
          name: 'Base',
          services: [{ description: 'A', price: 5 }],
          sound_fee: 1,
          travel_fee: 2,
          accommodation: null,
          discount: null,
          created_at: '',
          updated_at: '',
        },
      ],
    });
    const div = document.createElement('div');
    const root = createRoot(div);
    await act(async () => {
      root.render(
        <SendQuoteModal
          open
          onClose={() => {}}
          onSubmit={() => {}}
          artistId={2}
          clientId={3}
          bookingRequestId={4}
        />,
      );
    });
    const select = div.querySelector('select') as HTMLSelectElement;
    await act(async () => {
      select.value = '1';
      select.dispatchEvent(new Event('change', { bubbles: true }));
    });
    const summary = div.querySelector('div.text-sm.mt-2') as HTMLDivElement;
    expect(summary.textContent).toContain(formatCurrency(8));
    root.unmount();
  });
});
