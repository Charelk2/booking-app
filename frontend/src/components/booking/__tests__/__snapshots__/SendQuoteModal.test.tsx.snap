// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendQuoteModal matches snapshot 1`] = `
<div
  class="fixed inset-0 bg-black/40 flex items-center justify-center z-50"
>
  <div
    class="bg-white rounded-lg shadow-lg w-full max-w-md p-4"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <h2
        class="text-lg font-medium"
      >
        View Quote
      </h2>
    </div>
    <div
      class="space-y-3 max-h-[60vh] overflow-y-auto"
    >
      <div
        class="flex flex-col gap-y-2 mb-2 text-sm"
      >
        <div>
          Quote #2025-4394
        </div>
        <div>
          July 22nd, 2025
        </div>
        <input
          class="border rounded p-1"
          placeholder="Description"
          type="text"
          value=""
        />
      </div>
      <label
        class="flex items-center gap-2 text-sm font-normal mb-2 border rounded p-2"
        for="service-fee"
      >
        <span
          class="flex-1"
        >
          Live Performance
           fee
        </span>
        <input
          class="w-24 border rounded p-1 text-left focus:outline-none focus:ring-2 focus:ring-brand"
          id="service-fee"
          inputmode="numeric"
          placeholder="Enter amount"
          type="number"
          value="0"
        />
      </label>
      <label
        class="flex items-center gap-2 text-sm font-normal mb-2 border rounded p-2"
        for="sound-fee"
      >
        <span
          class="flex-1"
        >
          Sound fee
        </span>
        <input
          class="w-24 border rounded p-1 text-left focus:outline-none focus:ring-2 focus:ring-brand"
          id="sound-fee"
          inputmode="numeric"
          placeholder="Enter amount"
          type="number"
          value="0"
        />
      </label>
      <label
        class="flex items-center gap-2 text-sm font-normal mb-2 border rounded p-2"
        for="travel-fee"
      >
        <span
          class="flex-1"
        >
          Travel fee
        </span>
        <input
          class="w-24 border rounded p-1 text-left focus:outline-none focus:ring-2 focus:ring-brand"
          id="travel-fee"
          inputmode="numeric"
          placeholder="Enter amount"
          type="number"
          value="0"
        />
      </label>
      <button
        aria-busy="false"
        class="inline-flex items-center justify-center rounded-md font-semibold focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 transition-transform transition-colors active:scale-95 min-h-12 min-w-12 px-4 py-2 text-sm bg-white border border-gray-300 text-gray-800 hover:bg-gray-50 focus:ring-gray-300 text-sm"
        type="button"
      >
        <span
          class=""
        >
          Add Item
        </span>
      </button>
      <label
        class="flex flex-col text-sm font-normal"
        for="accommodation"
      >
        Accommodation (optional)
        <textarea
          class="w-full border rounded p-1 focus:outline-none focus:ring-2 focus:ring-brand"
          id="accommodation"
          placeholder="Optional: e.g. 500"
        />
      </label>
      <label
        class="flex items-center gap-2 text-sm font-normal mb-2"
        for="discount"
      >
        <span
          class="flex-1"
        >
          Discount (optional)
        </span>
        <input
          class="w-24 border rounded p-1 text-left focus:outline-none focus:ring-2 focus:ring-brand"
          id="discount"
          inputmode="numeric"
          placeholder="Optional: e.g. 500"
          type="number"
          value="0"
        />
      </label>
      <label
        class="flex flex-col text-sm font-normal"
        for="expires-hours"
      >
        Expires in
        <select
          class="w-full border rounded p-1 focus:outline-none focus:ring-2 focus:ring-brand"
          id="expires-hours"
        >
          <option
            value=""
          >
            No expiry
          </option>
          <option
            value="24"
          >
            1 day
          </option>
          <option
            value="72"
          >
            3 days
          </option>
          <option
            value="168"
          >
            7 days
          </option>
        </select>
      </label>
      <hr
        class="border-t my-2"
      />
      <div
        class="text-sm"
      >
        <span
          class="text-gray-500"
        >
          Subtotal: 
          R 0,00
        </span>
        <br />
        <span
          class="font-medium"
        >
          Total (after discount): 
          R 0,00
        </span>
      </div>
    </div>
    <div
      class="flex gap-2 mt-4"
    >
      <button
        aria-busy="false"
        class="inline-flex items-center justify-center rounded-md font-semibold focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 transition-transform transition-colors active:scale-95 min-h-12 min-w-12 px-4 py-2 text-sm bg-white border border-gray-300 text-gray-800 hover:bg-gray-50 focus:ring-gray-300"
        type="button"
      >
        <span
          class=""
        >
          Cancel
        </span>
      </button>
      <button
        aria-busy="false"
        class="inline-flex items-center justify-center rounded-md font-semibold focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 transition-transform transition-colors active:scale-95 min-h-12 min-w-12 px-4 py-2 text-sm bg-brand hover:bg-brand-dark text-white focus:ring-brand-dark"
        title="This quote will be sent to the client"
        type="button"
      >
        <span
          class=""
        >
          Submit Request
        </span>
      </button>
    </div>
  </div>
</div>
`;
